### PR TITLE
SCB-856 Implement reaction of the event in Alpha Server

### PR DESCRIPTION
**Pull request by cherrylzhao**
Implement reaction of TCC event in alpha server, which involves
-  rpc OnConnected (GrpcServiceConfig) returns (stream GrpcTccCordinateCommand)
- rpc participate(GrpcTccParticipateEvent) returns (GrpcAck)
- rpc OnTccTransactionEnded (GrpcTccTransactionEndedEvent) returns (GrpcAck)
- rpc OnDisconnected (GrpcServiceConfig) returns (GrpcAck)
- Add junit test framework code for alpha TCC workflow.
- Integrate alpha saga and tcc within same boot which are different with port.

### CODE REVIEW

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallback.java [hunk 1]
```diff
@@ -18,11 +18,10 @@
 package org.apache.servicecomb.saga.alpha.server.tcc;
 
 import org.apache.servicecomb.saga.alpha.server.tcc.event.ParticipatedEvent;
-import org.apache.servicecomb.saga.common.TransactionStatus;
 
 public interface OmegaCallback {
 
-  void compensate(ParticipatedEvent event, TransactionStatus status);
+  void invoke(ParticipatedEvent event, String status);
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallback.java]: Why did you change the status to String?

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallbacksRegistry.java [hunk 2]
```diff
@@ -32,6 +33,10 @@
 
   private final static Map<String, Map<String, OmegaCallback>> REGISTRY = new ConcurrentHashMap<>();
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallbacksRegistry.java]: Please don't use the static Map as the registry, it could introduce lots of trouble here.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallbacksRegistry.java [hunk 3]
```diff
@@ -50,9 +55,18 @@ public static void register(GrpcServiceConfig request, StreamObserver<GrpcTccCoo
    * @param serviceName service name
    * @param instanceId instance id
    * @return Grpc omega TCC callback
+   * @throws AlphaException trigger this exception while missing omega callback by service name
    */
-  public static OmegaCallback retrieve(String serviceName, String instanceId) {
-    return REGISTRY.getOrDefault(serviceName, emptyMap()).get(instanceId);
+  public static OmegaCallback retrieve(String serviceName, String instanceId) throws AlphaException {
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallbacksRegistry.java]: I don't think these method should be static.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/TransactionEventRegistry.java [hunk 4]
```diff
@@ -28,7 +28,7 @@
  */
 public final class TransactionEventRegistry {
 
-  private final static Map<String, List<ParticipatedEvent>> REGISTRY = new ConcurrentHashMap<>();
+  private final static Map<String, Set<ParticipatedEvent>> REGISTRY = new ConcurrentHashMap<>();
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/TransactionEventRegistry.java]: Same issue here.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallbacksRegistry.java (hunk 2 above)
Comment (cherrylzhao) [reply alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/OmegaCallbacksRegistry.java]: i think concurrent problem don't exist

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+  @Override
+  public void onTccTransactionEnded(GrpcTccTransactionEndedEvent request, StreamObserver<GrpcAck> responseObserver) {
+    try {
+      for (ParticipatedEvent event : TransactionEventRegistry.retrieve(request.getGlobalTxId())) {
+        OmegaCallbacksRegistry.retrieve(event.getServiceName(),
+            event.getInstanceId()).invoke(event, request.getStatus());
+      }
+    } catch (AlphaException ex) {
+      responseObserver.onNext(REJECT);
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java]: We need to write the log for the exception, otherwise the exception is eaten by us.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    }
+    responseObserver.onNext(ALLOW);
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java]: If there is AlphException throw, we still send the ALLOW message back to the omega.

File: alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/tcc/server/AlphaTccServerTest.java [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+
+  private GrpcTccTransactionEndedEvent newTxEnd(String status) {
+    return GrpcTccTransactionEndedEvent.newBuilder()
+        .setGlobalTxId(globalTxId)
+        .setLocalTxId(localTxId)
+        .setStatus(status)
+        .build();
+  }
+
+  private GrpcAck onCompensation(GrpcTccCoordinateCommand command) {
```
Comment (WillemJiang) [inline alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/tcc/server/AlphaTccServerTest.java]: It's not onCompensation.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/TccGrpcServerConfig.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+package org.apache.servicecomb.saga.alpha.server.tcc;
+
+import org.apache.servicecomb.saga.alpha.server.GrpcServerConfig;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnExpression;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConditionalOnExpression("'${alpha.mode:SAGA}'.contains('TCC')")
+public class TccGrpcServerConfig extends GrpcServerConfig {
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/TccGrpcServerConfig.java]: I'm not sure if the "alpha.mode:TCC" can do the work.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/TccGrpcServerConfig.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+  @Value("${alpha.server.tcc-port:8080}")
+  private int port;
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/TccGrpcServerConfig.java]: It could be great if we can let the saga and tcc service share the same port.  In this way, we don't need to open a new port for the new added protocol.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+  @Override
+  public void participate(GrpcTccParticipatedEvent request, StreamObserver<GrpcAck> responseObserver) {
+    TransactionEventRegistry.register(ParticipateEventFactory.create(request));
+    responseObserver.onNext(ALLOW);
+    responseObserver.onCompleted();
+  }
+
+  @Override
+  public void onTccTransactionEnded(GrpcTccTransactionEndedEvent request, StreamObserver<GrpcAck> responseObserver) {
+    responseObserver.onNext(tccCallbackEngine.execute(request) ? ALLOW : REJECT);
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java]: tccCallbackEngine may throw exception here.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+  private final TccCallbackEngine tccCallbackEngine;
+
+  public GrpcTccEventService(
+      TccCallbackEngine tccCallbackEngine) {
+    this.tccCallbackEngine = tccCallbackEngine;
+  }
+
+  @Override
+  public void onConnected(GrpcServiceConfig request, StreamObserver<GrpcTccCoordinateCommand> responseObserver) {
```
Comment (WillemJiang) [inline alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/GrpcTccEventService.java]: We need add some log for these calls.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/callback/OmegaCallbackWrapper.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+public class OmegaCallbackWrapper implements OmegaCallback {
+
+  @Override
+  public void invoke(ParticipatedEvent event, TransactionStatus status) {
+    OmegaCallback omegaCallback = OmegaCallbacksRegistry.retrieve(event.getServiceName(), event.getInstanceId());
+    try {
+      omegaCallback.invoke(event, status);
+    } catch (Exception ex) {
+      OmegaCallbacksRegistry.removeByValue(event.getServiceName(), omegaCallback);
```
[the remark this answers is not in the document]
Comment (cherrylzhao) [reply alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/callback/OmegaCallbackWrapper.java]: log has been written by TccCallbackEngine

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/callback/TccCallbackEngine.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+  @Override
+  public boolean execute(GrpcTccTransactionEndedEvent request) {
+    boolean result = true;
+    for (ParticipatedEvent event : TransactionEventRegistry.retrieve(request.getGlobalTxId())) {
+      try {
+        omegaCallbackWrapper.invoke(event, TransactionStatus.valueOf(request.getStatus()));
+      } catch (Exception ex) {
+        logError(event, ex);
+        result = false;
```
[the remark this answers is not in the document]
Comment (cherrylzhao) [reply alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/callback/TccCallbackEngine.java]: how about the check logic, all the participate will be called in foreach now.

File: alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/tcc/server/AlphaTccServerTest.java [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+
+  @AfterClass
+  public static void tearDown() {
+    clientChannel.shutdown();
+    clientChannel = null;
+  }
+
+  @Before
+  public void before() {
+    System.out.println(" globalTxId " + globalTxId);
```
[the remark this answers is not in the document]
Comment (cherrylzhao) [reply alpha/alpha-server/src/test/java/org/apache/servicecomb/saga/alpha/tcc/server/AlphaTccServerTest.java]: okey.

File: alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/callback/TccCallbackEngine.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (cherrylzhao) [reply alpha/alpha-server/src/main/java/org/apache/servicecomb/saga/alpha/server/tcc/callback/TccCallbackEngine.java]: result was returned in GrpcOmegaTccCallback, please check it.

**Comment by WillemJiang** [comment]
Note, the iCLA is received.
